File: impls/c++/src/step0_repl.cpp
#include <string>
#include <iostream>

std::string read(std::string input) {
  return input;
}
std::string eval(std::string input) {
  return input;
}
std::string print(std::string input) {
  return input;
}
std::string rep(std::string input) {
  return print(eval(read(input)));
}


int main() {
  std::string input;
  std::string prompt = "user> ";
  std::cout << prompt;
  while (std::getline(std::cin, input)) {
    auto output = rep(input);
    std::cout << output << std::endl;
    std::cout << prompt;
  }
  return 0;
}

File: impls/myjulia/step0_repl.jl

function READ(str)
    str
end

function EVAL(str)
    str
end

function PRINT(str)
    str
end

function rep(str)
    PRINT(EVAL(READ(str)))
end

function main()
    prompt = "user> "
    print(prompt)
    input = readline(keep=true)
    while input != ""
        print(rep(input))
        print(prompt)
        input = readline(keep=true)
    end
end

main()
